js/Makefile

```diff
@@ -201,16 +223,16 @@ android.log:
 .PHONY: android.emulator
 android.emulator:
 	$(call check-program, adb emulator)
-	$(eval ANDROID_EMULATOR ?= `emulator -list-avds | head -n1`)
+	$(eval ANDROID_EMULATOR ?= $(shell emulator -list-avds | head -n1))
 	@if [ -z "$(ANDROID_EMULATOR)" ]; then \
 	  >&2 echo "ERROR: no Android emulator AVD found"; exit 1; \
 	fi
-	emulator @$(ANDROID_EMULATOR) -logcat *:i
+	emulator @$(ANDROID_EMULATOR) -logcat *:i -no-snapshot-save
 
 .PHONY: android.reverse_tcp
 android.reverse_tcp:
 	$(call check-program, adb)
-	$(eval ANDROID_DEVICE ?= `adb devices | tail +2 | head -1 | cut -f 1`)
+	$(if $(ANDROID_DEVICE),,$(eval ANDROID_DEVICE = $(shell adb devices | tail +2 | head -1 | cut -f 1)))
```

Comment:
```suggestion
	$(if $(ANDROID_DEVICE),,$(eval ANDROID_DEVICE ?= $(shell adb devices | tail +2 | head -1 | cut -f 1)))
```

nope?

Reply:
I just tested and both syntaxes work: if the variable is set by the user, it is taken into account, otherwise this line lists the devices and sets the variable.

Reply:
Oh yes, sorry I was trying to fix your line but didn't get it entirely

If your change has the same effect we had previously, I strongly prefer the original way

Reply:
```suggestion
	$(eval ANDROID_DEVICE ?= `adb devices | tail +2 | head -1 | cut -f 1`)
```

Reply:
Tested this line, it doesn't work.
```
eval_test:
	$(eval ANDROID_DEVICE ?= `adb devices | tail +2 | head -1 | cut -f 1`)
	echo $(ANDROID_DEVICE)
```

output:
```
echo
```

And without the `?`:
```
eval_test:
	$(eval ANDROID_DEVICE ?= `adb devices | tail +2 | head -1 | cut -f 1`)
	echo $(ANDROID_DEVICE)
```

output:
```
echo `adb devices | tail +2 | head -1 | cut -f 1`
emulator-5554
```

But with the 2nd option (without `?`), it will ignore the value set by the user.

Reply:
ok, that's probably due do the `ANDROID_DEVICE ?=` at the beginning of the file

I dislike this complicated Makefile style but if it fixes real problem, up to you to merge like this :+1:

Reply:
fyi, my favorite method is just:

```make
ANDROID_DEVICE ?= `adb devices | tail +2 | head -1 | cut -f 1`
```

at the beginning of the file, it will not assign the result of `adb devices` but just the shell command to run only when it will really be called

as soon as we have `eval` in Makefile, we reduce the maintainability

an alternative I like is having only one helper function that centralizes all the complicated stuff, so you can just depend on that helper to do the eval(triple(condition(depending(on(100(lines(above)))))))` which makes it easier to understand at the end

Reply:
Ok got it but if we do this:
```
ANDROID_DEVICE ?= `adb devices | tail +2 | head -1 | cut -f 1`
```

We will overide the behavior of `react-native run-android` that automatically create and launch a temporary emulator (even if no AVD is available) in `android.debug` and `android.release` since `ANDROID_DEVICE` will always be set. I'll keep it as it is.